### DISCUSSION TITLE
Feature request: flock

### DISCUSSION THREAD

**Discussion by 7underlines** [vlang/v#11711] [Ideas]
Hi, does v have something like flock from PHP? https://www.php.net/manual/en/function.flock.php
Because I'm looking for a way to prevent double running a long process that is started via cron.
On discord someone mentioned that I could maybe use os.signal, but I'm not so sure how to implement it.

**Comment by dumblob** [comment]
1. your process shall create a temporary directory (because that is on most OSes atomical unlike creating a file)
2. if it failed (because the directory already existed), quit immediately
3. otherwise (i.e. the directory was successfully created) create a signal handler for SIGINT QUIT HUP ABRT TERM EXIT which deletes the created temporary directory
4. add `defer{ remove_my_lock_directory() }` in `main()`
5. nothing else, that's it :wink:

**Comment by 7underlines** [comment]
@dumblob Thank you! It's working so far. I've implemented it like this:

```
fn on_signal() {
	flock_dir := [os.temp_dir(), '.myapp_custom_flock_processing'].join(os.path_separator)
	os.rmdir(flock_dir)
	exit(1)
}

fn main() {
	flock_dir := [os.temp_dir(), '.myapp_custom_flock_processing'].join(os.path_separator)
	os.mkdir(flock_dir) or {exit(0)}

	os.signal(1, on_signal)
	os.signal(2, on_signal)
	os.signal(3, on_signal)
	os.signal(6, on_signal)
	os.signal(15, on_signal)

	// here I do stuff

	os.rmdir(flock_dir)
}
```

**Comment by dumblob** [comment]
@logTom glad it works for you. And oh, I forgot to remove the dir at return from `main()`. I've edited my suggestion above :wink:.

**Comment by 7underlines** [comment]
Ah, there is one caveat - this doesn't work if v panics or I stop the program midway with an exit(1) for debugging. I also tried it with defer.

**Comment by dumblob** [comment]
>Ah, there is one caveat - this doesn't work if v panics or I stop the program midway with an exit(1) for debugging. I also tried it with defer.

I'm glad you shout here about this unwise decision. I'm shouting as well. We'll all cry one day about this IMHO biggest issue of *V*. Read more about `panic()` being the most harmful feature in *V*: https://github.com/vlang/v/issues/2030#issuecomment-532746482 (IMHO this antifeature has the potential to diminish widespread use of *V*).

On the positive side, `defer{}` section shall be called irrespectively of `panic()` if I'm not mistaken. So you're observing a buggy behavior IMHO.

@medvednikov and others, would you "dare" to contemplate about `panic()` being more and more harmful? In light of channels being currently implemented into the language this'll become a hot topic some day sooner or later anyway.
